Approved/3-multidimensional-arrays.md

```diff
@@ -0,0 +1,747 @@
+- `internal function NDArrayOffsetUnsafe<'TArray>(data : 'TArray) : Int`
+
+Each of these five functions would be `body intrinsic;`, and together form the contract between the runtime and the core Q# libraries needed to support this proposed feature (see Example 7, below). By necessity, these functions are "unsafe," in that direct access to these functions would allow violating invariants of multidimensional arrays or bypass the type system to expose runtime failures, necessitating the `internal` modifier. In particular, these functions may be given access to elements that do not exist insofar as Q# user code is concerned (i.e.: are not accessible given the current strides, shapes, and offsets); these elements may have reference counts of zero, such that referring to these elements results in undefined behavior. It is the responsibility of the Q# foundation, core, and standard libraries to ensure that undefined behavior is not exposed to user code.
+
+In user code, multidimensional arrays can be created and manipulated using Q# standard library functions and operations built up using the above internal functions, as described in example 7 below.
+For instance, the "diagonal" elements of a 2-D array can be extracted into a 1-D array using `Diagonal2<'T>(array : [|'T|]) : ['T]`, as shown in example 8, below.
+Please see https://github.com/microsoft/QuantumLibraries/issues for more details about library functions for working with multidimensional arrays.
+
+#### Examples
+
+Example 1:
+Declaring and indexing into variables of type `[|Double|]` and `[||Int||]` using literals.
+
+```qsharp
+let z = [
+    // Inside [] delimiters, | | delimiters refer not to array expressions, but
+    // to "rows" of the two-dimensional array literal.
+    | 1.0, 0.0 |,
+    | 0.0, -1.0 |
+];
+Message($"{z[0, 1]}"); // ← 0.0
+
+// It is a compile-time error for the "rows" denoted by | ... | to be uneven:
+let incorrect = [
+    | 1.0, 2.0 |,
+    | 10.0, 20.0, 30.0 | // ← error, since this would declare a "ragged" array
+];
+
+let data = [
+    // We can nest two levels of | ... | to get three-dimensional arrays.
+    |
+        |0, 1|,
+        |2, 3|
+    |,
+
+    |
+        |4, 5|,
+        |6, 7|
+    |
+];
+Message($"{data[0, 1, 0]}"); // ← 6
+```
+
+Example 2:
+Mixing 1D and multidimensional arrays.
+
+```qsharp
+// Here, we declare the first two levels of indexing as being rectangular,
+// resulting in a three-dimensional array.
+let data = [
+    |
+        // After rows, [] denote arrays as elements again.
+        | [0], [1, 2] |,
+        | [3, 4, 5], [6, 7, 8, 9] |
+    |,
+
+    |
+        | [10, 11, 12, 13, 14], [15, 16, 17, 18, 19, 20] |,
+        | [21, 22, 23, 24, 25, 26, 27], [28, 29, 30, 31, 32, 33, 34, 35] |
+    |
+];
+// data: [||[Int]||] (that is, a three-dimensional array of arrays of integers)
+// Get the zero-th "plane," first "row", zeroth "column," and third element.
+Message($"{data[0, 1, 0][2]}"); // ← 5
+```
+
+Example 3:
```

Comment:
@cgranade What exactly is the signature of Concatenated2? 
It is hard to follow the example without the signature of Concatenated2, and I am not sure the signature that the example implies is indeed the one you intended. Could you please revise that example?